### PR TITLE
Update CLI to use feature flags, instead of project calls

### PR DESCRIPTION
**Pull request by stefanmcshane**
## Issue ID

N/A

## What does this PR do?

- Removed all project-gets, and start passing around feature flags where necessary
- Evaluate flags passed in by CLI as Viper was skipping them due to a lack of global initialisation. This is an intermediary solution for reading flag values at runtime
- Remove token from porter config on failed auth login with `--token` flag

### CODE REVIEW

File: api/client/api.go [hunk 1]
```diff
@@ -82,33 +80,6 @@ func NewClientWithConfig(ctx context.Context, input NewClientInput) (Client, err
 // ErrNoAuthCredential returns an error when no auth credentials have been provided such as cookies or tokens
 var ErrNoAuthCredential = errors.New("unable to create an API session with cookie nor token")
 
-// NewClient constructs a new client based on a set of options
```
Comment (stefanmcshane) [inline api/client/api.go:85]: Housekeeping - unused function

File: cli/cmd/config/config.go [hunk 2]
```diff
@@ -56,6 +62,11 @@ func initAndLoadConfig() (CLIConfig, error) {
 	viper.SetConfigType("yaml")
 	viper.AddConfigPath(porterDir)
 
+	err = createAndLoadPorterYaml(porterDir)
```
Comment (stefanmcshane) [inline cli/cmd/config/config.go:65]: Evaluate porter.yaml before flags and environment variables

File: cli/cmd/commands/all.go [hunk 3]
```diff
@@ -50,3 +50,69 @@ func RegisterCommands() (*cobra.Command, error) {
 	rootCmd.AddCommand(registerCommand_Version(cliConf))
 	return rootCmd, nil
 }
+
+// overrideConfigWithFlags grabs the runtime value of registered flags, and overrides the values in CLIConfig.
+// It was done this was to reduce the size of a refactor, as the codebase conflates initialisation of the commands, with the runtime values.
```
Comment (ferozemohideen) [inline cli/cmd/commands/all.go:55]: nit: comment missing a word?

File: cli/cmd/commands/all.go [hunk 4]
```diff
@@ -50,3 +50,69 @@ func RegisterCommands() (*cobra.Command, error) {
...
+		{
+			uintName:         "helm_repo",
+			uintConfigTarget: &config.HelmRepo,
+		},
+	}
+	for _, fl := range flagsToOverride {
+		if fl.stringName != "" {
+			st, _ := cmd.Flags().GetString(fl.stringName)
+			if st != "" {
+				*fl.stringConfigTarget = st
```
Comment (ferozemohideen) [inline cli/cmd/commands/all.go:107]: are these pointer operations panic safe?

**Review by ferozemohideen — APPROVED** [review]

File: cli/cmd/commands/auth.go [hunk 5]
```diff
@@ -53,8 +60,9 @@ func registerCommand_Auth(cliConf config.CLIConfig) *cobra.Command {
 		Use:   "logout",
 		Short: "Logs a user out of a given Porter server",
 		Run: func(cmd *cobra.Command, args []string) {
-			err := checkLoginAndRunWithConfig(cmd.Context(), cliConf, args, logout)
+			err := checkLoginAndRunWithConfig(cmd, cliConf, args, logout)
 			if err != nil {
+				_ = cliConf.SetToken("")
```
Comment (ianedwards) [inline cli/cmd/commands/auth.go:65]: Would it make sense to clear project / cluster as well? Or do we set those on login

File: cli/cmd/config/config.go [hunk 6]
```diff
@@ -56,6 +59,11 @@ func initAndLoadConfig() (CLIConfig, error) {
...
```
Comment (ianedwards) [inline cli/cmd/config/config.go:62]: Is this referring to the file where configuration is stored?

**Review by ianedwards — APPROVED** [review]

File: cli/cmd/commands/all.go (hunk 4 above)
Comment (stefanmcshane) [reply cli/cmd/commands/all.go:107]: yeah - they are only used in this one function as an easier way to read the command line values in for now. There are much much better ways to manage this, but that seemed outside the scope of the PR for now as its big enough

File: cli/cmd/config/config.go (hunk 6 above)
Comment (stefanmcshane) [reply cli/cmd/config/config.go:62]: It is - welcome to viper. It stores an internal global state and I plan to remove it. The project that cobra and viper were created for (hugo) no longer use viper, and use a stripped down version of cobra

File: cli/cmd/commands/auth.go (hunk 5 above)
Comment (stefanmcshane) [reply cli/cmd/commands/auth.go:65]: They get set on login

**Pull request merged**